### PR TITLE
[gpu] NFC: Move MMA schedule deduction to Common/GPU/

### PR DESCRIPTION
**Pull request by antiagainst**
This prepares it to be shared by the LLVMGPU path.

### CODE REVIEW

**Comment by antiagainst** [comment]
The first one is purely copy over. The second commit actually updates and rewires the code.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+namespace mlir::iree_compiler {
+
+/// Struct containing information about a matmul's shape and type.
+struct GPUMatmulShapeType {
+  int64_t mSize;
+  int64_t nSize;
+  int64_t kSize;
+  Type aType;
+  Type bType;
+  Type cType;
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h:18]: Thinking out loud, we might need to rethink this for integer types at some point, e.g. `V_WMMA_I32_16X16X16_IU8` for rdna3: https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  int64_t nSize;      // Native MMA size along N dimension
+  int64_t kSize;      // Native MMA size along K dimension
+  int64_t mWarpCount; // Number of subgroups along M dimension
+  int64_t nWarpCount; // Number of subgroups along N dimension
+  int64_t mTileCount; // Number of tiles per subgroup along M dimension
+  int64_t nTileCount; // Number of tiles per subgroup along N dimension
+  int64_t kTileCount; // Number of tiles along K dimension
+};
+
+/// Returns a schedule for using one of th the given MMA |intrinsics| to target
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h]:
```suggestion
/// Returns a schedule for using one of the given MMA |intrinsics| to target
```

**Review by qedawkins — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h (hunk 2 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h]: Fine to leave this as is and fix with an upcoming PR.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h (hunk 1 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h:18]: I feel the trend is actually more moving towards fp8 or other smaller floating point numbers.. Though integer quantization will remain for sure. People are just circling through these mechanisms anyway; just a matter of time. Will see.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.h:18]: yeah lots of crazy math and hardware when it comes to quantization, and it seems like people want to push a lot of the burden on to the compiler to handle their novel ideas/hardware well...

**Pull request merged**